functions/go/set-namespace-v2/namespace_transformer.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/krmfn"
+)
+
+
+const (
+	fnConfigAPIVersion = "fn.kpt.dev/v1alpha1"
+	legacyFnConfigKind = "SetNamespaceConfig"
+	fnConfigKind       = "SetNamespace"
+	dependsOnAnnotation    = "config.kubernetes.io/depends-on"
+	groupIdx     = 0
+	namespaceIdx = 2
+	kindIdx      = 3
+	nameIdx      = 4
+)
+
+
+var (
+	// <group>/namespaces/<namespace>/<kind>/<name>
+	namespacedResourcePattern = regexp.MustCompile(`\A([-.\w]*)/namespaces/([-.\w]*)/([-.\w]*)/([-.\w]*)\z`)
+	dependsOnKey = func(apiVersion, kind, name string) string {
+		group, _ := krmfn.ParseGroupVersion(apiVersion)
+		return fmt.Sprintf("%s/%s/%s", group, kind, name)
+	}
+)
+
+type Subject struct {
+	Kind string  `json:"kind,omitempty" yaml:"kind,omitempty"`
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+	ApiGroup string `json:"apiGroup,omitempty" yaml:"apiGroup,omitempty"`
+	Namespace string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
+}
+
+type NamespaceTransformer struct {
+	Namespace string
+	DependsOnMap map[string]bool
+	Errors []string
+}
+
+func (p *NamespaceTransformer) Config(o *krmfn.KubeObject) error {
+	switch {
+	case o.APIVersion() == "v1" && o.Kind() == "ConfigMap" && o.Name() == "kptfile.kpt.dev":
+		p.Namespace = o.GetStringOrDie("data", "name")
```

Comment:
`GetString` returns a potential error while `GetStringOrDie` exits on error. `***OrDie` pattern comes from kustomize, there are many usages in its code base.
If we want to be more conservative, we should use `GetString`. If we want to to more concise (fewer LoC), we can use `GetStringOrDie`.
We need to decide which one should we use here.

Reply:
Personally prefer not to handle the errors on the caller side (more clean code and let SDK decides how to reveal the message and give suggestions). I added the panic recovery for `***OrDie` pattern to give users some guidance on which resource's which field path causes the problem. https://github.com/GoogleContainerTools/kpt-functions-sdk/pull/531  
For example, I changed my `ClusterRoleBinding` resource `subjects` to be a list of string (it should be a list of new MapVariant), the output is
```bash
[RUNNING] "./set-ns"
[FAIL] "./set-ns" in 700ms
  Stderr:
    "2022/03/15 22:34:07 Resource(apiVersion=v1, kind=ClusterRoleBinding, Name=yuyu2) has unmatched field type: `subjects`"
    ""
  Exit code: 1
```